### PR TITLE
Helper keyword for iterating local work items

### PR DESCRIPTION
**Pull request by cmin764**
Add `@{results} =     For Each Input Work Item    <keyword>    <params>` keyword for consuming all work items in the local queue while applying a keyword over them.

Example
```robotframework
*** Settings ***
Documentation    Testing the helper keyword on consuming all work items in the queue.
Library          RPA.Robocorp.WorkItems
Library          OperatingSystem
Library          Collections


*** Variables ***
${idx}    ${1}


*** Keywords ***
Add Work Item
    [Arguments]  ${counter}

    # Get files from the current input work item.
    ${path} =        Get Work Item File    orders.txt    path=${OUTPUT_DIR}${/}orders${counter}.txt
    Log To Console    Getting: ${path}
    ${content} =     Get File    ${path}

    # Save a new file with the same content into one adjacent output work item.
    Create Output Work Item
    Set work item variables    name=CustomName${counter}   id=${counter}
    Add Work Item File    devdata/orders${counter}.txt
    Save Work Item
    [Return]    ${content}

Add Work Item Wrapper
    ${ret} =    Add Work Item    ${idx}
    Set Global Variable    ${idx}    ${idx + 1}

    # Any of these would fail if uncommented
    # Get Input Work Item
    # @{results} =     For Each Input Work Item    Add Work Item Wrapper
    
    [Return]    ${ret}


*** Tasks ***
Consume queue
    @{results} =     For Each Input Work Item    Add Work Item Wrapper
    Log List    ${results}
```

For the whole robot content including the prebuilt wheel with the changes in this PR see the attached archive. (make sure you rebuild your `rcc` env by removing the old one)

Latest robot (**4 Oct**): [work-items-matti.zip](https://github.com/robocorp/rpaframework/files/7278234/work-items-matti.zip)

_env.json_
```JSON
{
    "RPA_WORKITEMS_ADAPTER": "RPA.Robocorp.WorkItems.FileAdapter",
    "RPA_INPUT_WORKITEM_PATH": "%{ROBOT_ROOT}/devdata/items-in.json",
    "RPA_OUTPUT_WORKITEM_PATH": "%{ROBOT_ROOT}/output/items-out.json"
}
```

Look into the attached archive for the rest of the required files.

If it works, you should see some console output like so:
```console
Consume queue                                                         Getting: /Users/cmin/Robots/work-items-matti/output/orders1.txt
Getting: /Users/cmin/Robots/work-items-matti/output/orders2.txt
Getting: /Users/cmin/Robots/work-items-matti/output/orders3.txt
| PASS |
```

### Notes
- [Bugs](https://linear.app/robocorp/issue/LIB-12/fix-looping-of-work-items-bug) fixed
- While iterating work items already you can't (inside the iteration):
    - Iterate work items with the helper
    - Pull new input work items
    - Do it in the cloud (for now)

### ToDo
- [x] Confirmation from @janipalsamaki that this keyword works as expected
- [x] Unit and robot tests (adding/removing files under input/output work items, the helper keyword)
- [x] Review, refactor/accept,  release notes, squash-merge

### CODE REVIEW

**Review by mikahanninen — COMMENTED** [review]
interesting keyword, need to test myself before full approval

File: packages/main/src/RPA/Robocorp/WorkItems.py
Comment (mikahanninen) [inline packages/main/src/RPA/Robocorp/WorkItems.py:1292]: I guess that this row will fail on pure Python run

File: packages/main/src/RPA/Robocorp/WorkItems.py
Comment (mikahanninen) [inline packages/main/src/RPA/Robocorp/WorkItems.py:1282]: Add parameter descriptions and example how the keyword is used

**Review by fabioz — COMMENTED** [review]
Hi @cmin764 I just did a quick review and ended up finding some nitpicks on general python usage, not real bugs... feel free to ignore if you don't think it's pertinent.

Some unrelated things which I noticed (which seem more related to the library, so, just raising the discussion and may not be pertinent to the actual pull request):

- Shouldn't there be Robot Framework tests accompanying the library? (it'd be nice to see in tests how the library is actually used)
- Is the library meant to be used without Robot Framework or only with Python? (some parts of the code seem to work only with Robot Framework...)

**Comment by janipalsamaki** [comment]
I'm failing to figure out how to get access to the current work item (and its payload) in the callback. Any examples? :D

File: packages/main/src/RPA/Robocorp/WorkItems.py
Comment (cmin764) [reply packages/main/src/RPA/Robocorp/WorkItems.py]: I added now the possibility of providing a Py function instead of just a RF keyword. Is that what we'd expect?

**Comment by cmin764** [comment]
> * Framework tests accompanying the library? (it'd be nice to see in tests how the library is actually used)
> * Is the library meant to be used without Robot Framework or only with Python? (some parts of the code seem to work only with Robot Framework...)

Thanks for your attention to details.
- The library should work both under Py and RF, will take care of that.
- Yes, definitely I'll add both unit and robot tests. Just wanted the reviewers to have a glance at the logic first (thus marking the PR as draft) ahead of tests since is much faster to change code at this stage.

Will take care of your notes in this PR. (or another one with a Linear placeholder)

File: packages/main/src/RPA/Robocorp/WorkItems.py
Comment (cmin764) [reply packages/main/src/RPA/Robocorp/WorkItems.py:346]: Yes, that's an odd piece of code, refactored.

**Comment by cmin764** [comment]
> I'm failing to figure out how to get access to the current work item (and its payload) in the callback. Any examples? :D

There's one example right in the PR [description](https://github.com/robocorp/rpaframework/pull/241#issue-746944016) -> check this part:
```robotframework
${path} =        Get Work Item File    orders.txt    path=${OUTPUT_DIR}${/}orders${counter}.txt
```
and here's another example:
```robotframework
Log Payload
    ${payload} =     Get Work Item Payload
    Log To Console    ${payload}
    ${len} =     Get Length    ${payload}
    [Return]    ${len}

Log Payloads
    @{results} =     For Each Input Work Item    Log Payload
    Log   Items keys length: @{results}
```
where the helper automatically loads the next work item in the queue and afterwards you can load its payload in the mapping keyword, like I'm doing above. This example can be found in the keyword docstring too now.

----
Tested with this wheel: [rpaframework-11.3.0-py3-none-any.whl.zip](https://github.com/robocorp/rpaframework/files/7278232/rpaframework-11.3.0-py3-none-any.whl.zip)
 
And here's the entire robot I used (including examples above and more): [work-items-matti.zip](https://github.com/robocorp/rpaframework/files/7278234/work-items-matti.zip)

**Comment by cmin764** [comment]
### Caveats

- Using the `For Each Input Work Item` twice will re-pick the last item in the queue for executing the passed keyword over (and will happen the same each time you call it) -> might be solved in a polishing PR somewhere next week if necessary
- Providing explicitly `RPA_INPUT/OUTPUT...` env vars in the _env.json_ file will override the settings in the VSCode extension, thus not working as expected when picking up the next input work item for the robot run -> will be discussed in the extension call and fixed in the polishing PR

File: packages/main/src/RPA/Robocorp/WorkItems.py
Comment (mmokko) [inline packages/main/src/RPA/Robocorp/WorkItems.py:369]: Even easier to read code would be to have _save_input_to_disk / _save_output_to_disk and the do the file operation on the _save_to_disk function, would save a condition check, but that is matter of opinion of course :), anyway better this way than before. Just wanted to give any feedback :).

**Review by mmokko — APPROVED** [review]

File: packages/main/src/RPA/Robocorp/WorkItems.py
Comment (cmin764) [reply packages/main/src/RPA/Robocorp/WorkItems.py:65]: Every argument after `*` is necessary to be passed as a keyword argument when the function is called, without providing a default value for them (as in traditional keyword arguments). I added this here just to make sure the dev doesn't get lost in those 4 arguments the method receives.

File: packages/main/src/RPA/Robocorp/WorkItems.py
Comment (cmin764) [reply packages/main/src/RPA/Robocorp/WorkItems.py:369]: Yes, I like this separation you suggest! Too bad I don't see how can I simplify the code within the current infrastructure. The culprit starts from this:
```python
    def _get_item(self, item_id: str) -> Tuple[str, Dict[str, Any]]:
        idx = int(item_id)
        if idx < len(self.inputs):
            return "input", self.inputs[idx]

        if idx < (len(self.inputs) + len(self.outputs)):
            return "output", self.outputs[idx - len(self.inputs)]

        raise ValueError(f"Unknown work item ID: {item_id}")
```
So solving that first might lead to a better codebase allowing this proposed refactor to happen.

**Pull request merged**
